scripts/migrate-auth0-users.js

```diff
@@ -78,107 +155,97 @@ async function main () {
-                // of the time this was caused by the user already existing
-                console.error(`created user failed for ${user.email}`)
-                console.error(json)
-              } else {
-                // user creation successful!
-                console.log(`created user ${user.email}`)
-              }
-            })
-            .catch(console.error)
+    const json = await response.json()
+
+    // update the progress of fetching users
+    numUsersFetched += json.users.length
+    console.log(`fetched ${numUsersFetched} users so far`)
+    totalUsers = json.total
+    page++
+
+    // iterate through fetched users
+    for (let user of json.users) {
+      // only users that have a matching clientId should be created in the
+      // new tenant, so assume they shouldn't be created until finding a
+      // match
+      let shouldCreateInNewApp = false
+
+      // replace all user and app metadata with only the matching from
+      // client data and then replace the client id with the to client id
+      const metadatas = ['user_metadata', 'app_metadata']
+      metadatas.forEach(metadataKey => {
+        if (!user[metadataKey]) return
+        const metadata = user[metadataKey]
+        if (!metadata.datatools) return
+        metadata.datatools = metadata.datatools.filter(
+          dt => dt.client_id === from.clientId
+        )
+        if (metadata.datatools.length > 0) {
+          shouldCreateInNewApp = true
+          metadata.datatools[0].client_id = to.clientId
         }
       })
+
+      // no match found, continue to the next user in the current batch
+      if (!shouldCreateInNewApp) continue
+
+      // can't copy over passwords, so set the password of the copied user
+      // as a UUID so that it's extremely unlikely for some random person
+      // to guess the password. The user will have to reset their password
+      // to gain access.
+      user.password = uuid()
+
+      // add required connection name (this is the default value created
+      // by Auth0, so it could theoretically be different.)
+      user.connection = DEFAULT_CONNECTION_NAME
+
+      // omit items that will cause the creation to fail or cause weirdness
+      // in new data values
+      user = omit(
+        user,
+        [
+          'logins_count',
+          'last_ip',
+          'last_login',
+          'last_password_reset',
+          'created_at',
+          'updated_at',
+          'identities'
+        ]
+      )
+
+      // modify the user_id so an extra prefix isn't created
+      user.user_id = user.user_id.substr(6)
+
+      // attempt to create the user
+      const createUserResult = await createUser(to.domain, toToken, user)
+      if (!createUserResult.success) {
+        // creation of user not successful, check if script should try again by
+        // deleting an existing user
+        if (
+          createUserResult.responseJson.message === 'The user already exists.' &&
```

Comment:
Is there a more stable way to check for this? I'm guessing not, since the HTTP status code is probably also used in other cases, but just wanted to check.

Reply:
Yeah, not really. I think this message was the most specific part of the response.